src/kdf.c

```diff
@@ -124,49 +241,245 @@ static int p11prov_hkdf_derive(void *ctx, unsigned char *key, size_t keylen,
+        .prfHashMechanism = hkdfctx->hash_mech,
+        .ulSaltType = 0,
+        .pSalt = NULL,
+        .ulSaltLen = 0,
+        .hSaltKey = CK_INVALID_HANDLE,
+    };
+    CK_MECHANISM mechanism = {
+        .mechanism = hkdfctx->mechtype,
+        .pParameter = &params,
+        .ulParameterLen = sizeof(params),
+    };
+    uint8_t info[2 + 1 + 255 + 1 + 255];
+    size_t infolen, i;
+    uint16_t keysize;
+    CK_RV ret;
+
+    P11PROV_debug(
+        "tls13 expand label (prefix:%p[%zu], label:%p[%zu], data:%p[%zu])",
+        prefix, prefixlen, label, labellen, data, datalen);
+
+    if (prefix == NULL || prefixlen == 0 || label == NULL || labellen == 0
+        || (prefixlen + labellen > 255) || (datalen > 0 && data == NULL)
+        || (datalen == 0 && data != NULL) || (datalen > 255)
+        || (keylen > 65535)) {
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    /* ref: RFC 8446 - 7.1 Key Schedule */
+    infolen = 2 + 1 + prefixlen + labellen + 1 + datalen;
+    i = 0;
+    keysize = htobe16(keylen);
+    memcpy(&info[i], &keysize, sizeof(uint16_t));
+    i += sizeof(uint16_t);
+    info[i] = (uint8_t)(prefixlen + labellen);
+    i += 1;
+    memcpy(&info[i], prefix, prefixlen);
+    i += prefixlen;
+    memcpy(&info[i], label, labellen);
+    i += labellen;
+    info[i] = (uint8_t)(datalen);
+    i += 1;
+    if (datalen > 0) {
+        memcpy(&info[i], data, datalen);
+        i += datalen;
     }
+    if (infolen != i) {
+        OPENSSL_cleanse(&info, i);
+        return CKR_HOST_MEMORY;
+    }
+
+    params.pInfo = info;
+    params.ulInfoLen = infolen;
 
-    ret = p11prov_derive_key(hkdfctx->provctx, slotid, &mechanism, pkey_handle,
-                             key_template, 5, &hkdfctx->session, &dkey_handle);
+    ret = inner_derive_key(hkdfctx->provctx, keyobj, &hkdfctx->session,
+                           &mechanism, keylen, dkey_handle);
+
+    OPENSSL_cleanse(&info, infolen);
+    return ret;
+}
+
+static CK_RV p11prov_tls13_derive_secret(P11PROV_KDF_CTX *hkdfctx,
+                                         P11PROV_OBJ *keyobj, size_t keylen,
+                                         CK_OBJECT_HANDLE *dkey_handle)
+{
+    P11PROV_OBJ *zerokey = NULL;
+    CK_HKDF_PARAMS params = {
+        .bExtract = CK_TRUE,
+        .bExpand = CK_FALSE,
+        .prfHashMechanism = hkdfctx->hash_mech,
+        .ulSaltType = CKF_HKDF_SALT_DATA,
+        .hSaltKey = CK_INVALID_HANDLE,
+        .pInfo = NULL,
+        .ulInfoLen = 0,
+    };
+    CK_MECHANISM mechanism = {
+        .mechanism = CKM_HKDF_DATA,
+        .pParameter = &params,
+        .ulParameterLen = sizeof(params),
+    };
+    uint8_t saltbuf[EVP_MAX_MD_SIZE] = { 0 };
+    uint8_t zerobuf[EVP_MAX_MD_SIZE] = { 0 };
+    size_t saltlen;
+    size_t hashlen;
+    CK_RV ret;
+
+    ret = p11prov_digest_get_digest_size(hkdfctx->hash_mech, &hashlen);
     if (ret != CKR_OK) {
+        return ret;
+    }
+    saltlen = hashlen;
+
+    if (hkdfctx->salt) {
```

Comment:
Style: this block deserves a separate func

Reply:
FYI: we generally do not split out a func unless there is more than one caller.